core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpService.java

```diff
@@ -232,19 +235,48 @@ public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exc
             case HTTP_RESPONSE:
-                return f.thenApply(
-                        msg -> new ExceptionFilteredHttpResponse(ctx, req, (HttpResponse) invoke(ctx, req, msg),
-                                                                 exceptionHandler));
+                if (useBlocking) {
+                    return f.thenApplyAsync(
+                            msg -> new ExceptionFilteredHttpResponse(ctx, req,
+                                                                     (HttpResponse) invoke(ctx, req, msg),
+                                                                     exceptionHandler),
+                            ctx.blockingTaskExecutor());
+                } else {
+                    return f.thenApply(
+                            msg -> new ExceptionFilteredHttpResponse(ctx, req,
+                                                                     (HttpResponse) invoke(ctx, req, msg),
+                                                                     exceptionHandler));
+                }
+
             case COMPLETION_STAGE:
-                return f.thenCompose(msg -> toCompletionStage(invoke(ctx, req, msg)))
-                        .handle((result, cause) -> cause == null ? convertResponse(ctx, req, null, result,
-                                                                                   HttpHeaders.of())
-                                                                 : exceptionHandler.handleException(ctx, req,
-                                                                                                    cause));
+                if (useBlocking) {
+                    return f.thenComposeAsync(msg -> toCompletionStage(invoke(ctx, req, msg)),
+                                              ctx.blockingTaskExecutor())
+                            .handle((result, cause) ->
+                                            cause == null ? convertResponse(ctx, req, null, result,
+                                                                            HttpHeaders.of())
+                                                          : exceptionHandler.handleException(ctx, req, cause));
+                } else {
+                    return f.thenCompose(msg -> toCompletionStage(invoke(ctx, req, msg)))
+                            .handle((result, cause) ->
+                                            cause == null ? convertResponse(ctx, req, null, result,
+                                                                            HttpHeaders.of())
+                                                          : exceptionHandler.handleException(ctx, req, cause));
+                }
+
             default:
-                return f.thenApplyAsync(msg -> convertResponse(ctx, req, null, invoke(ctx, req, msg),
-                                                               HttpHeaders.of()),
-                                        ctx.blockingTaskExecutor());
+                if (useBlocking) {
```

Comment:
If a user uses `@Blocking`, he/she always sees this warning even though he/she uses it correctly. How about adding a flag for that in `Flags`?
```java
if (useBlocking) {
    return f.thenApplyAsync(
            msg -> convertResponse(ctx, req, null, invoke(ctx, req, msg),
                                   HttpHeaders.of()),
            ctx.blockingTaskExecutor());
} else {
    if (!USE_EVENT_LOOP_FOR_ANNOTATED_SERVICE) { // default is false
        logger.warn(...);
        return f.thenApplyAsync(
                msg -> convertResponse(ctx, req, null, invoke(ctx, req, msg),
                                       HttpHeaders.of()),
                ctx.blockingTaskExecutor());
    } else {
        return f.thenApply(msg -> convertResponse(ctx, req, null, invoke(ctx, req, msg),
                                                  HttpHeaders.of()));
    }
}
```

Reply:
Fixed 1efe13d

Reply:
@heowc Thanks for the update.
I think it's better to just remove the flag. So let's just make a breaking change here. I'm sorry about that. 😄 
We have talked about the migration plan again, but we thought it's too much and we can just leave the information about that in the release note well.